Add recommended operator health status (#6315)
operator-framework/operator-sdk: website/content/en/docs/best-practices/observability-best-practices.md

 - `Gauge` Value can be increased and decreased as needed.
5. `_total` suffix should be used for accumulating count. If your metrics has labels with high cardinality, like `pod`/`container` it usually means that you can aggregate it more, thus it will not require `_total` suffix.

#### Recommended Metrics

* `operator_health_status` - The metric reports the operator health status. It should be reported by the operator itself based on the internal operator validations.

  This metric allows the cluster users to view all operators health with a single metric and improves the overall cluster observability.  

>   Valid health status values are:

Reviewer: @sradco I found the comment you made [here](https://github.com/operator-framework/operator-sdk/pull/6315/files#r1107769855) useful. It would be nice to mention the reasoning behind having them as numbers in the doc too. I would suggest something like:

```
We would suggest to use numbers to represent the statuses in the values, since string values like: Healthy/Warning/Error, can be implemented in many ways that would cause unexpected consistency issues and disagreements, like spelling errors, capital letters etc.

One of the suggested set of values are:
* `0` - Indicates that the operator is healthy and working as expected.
* `1` - Indicates that the operator has some issues that needs to be addressed and can lead to loss of functionality.
* `2` - Indicates that the operator is unhealthy and there is a loss of functionality that should be addressed.

....

The set of values and their definitions should be clearly documented by the operator authors for any external observer or tool to infer them correctly or take necessary action accordingly. 
```

Author: Update. Please review again.

  * `0` - Indicates that the operator is healthy and working as expected.
  * `1` - Indicates that the operator has some issues that needs to be addressed and can lead to loss of functionality.
  * `2` - Indicates that the operator is unhealthy and there is a loss of functionality that should be addressed.

  Metric labels:

>   * `name` - The operator name.

Reviewer 2: Could you point to precedents for the name and format and/or provide explanations on why they are a good choice?

Quickly looking for similar things, I found:
~~~
name
    kube_node_status_condition
condition
    DiskPressure
    PIDPressure
    MemoryPressure
    Ready
value
    0/1
status
    false/true/unknown
~~~
https://github.com/kubernetes/kube-state-metrics/blob/main/docs/node-metrics.md

~~~
name
    kube_deployment_status_condition
condition
    Available
    Progressing
    ReplicaFailure
value
    0/1
~~~
https://github.com/kubernetes/kube-state-metrics/blob/main/docs/deployment-metrics.md

~~~
name
	cluster_operator_conditions
condition
	Available
	Degraded
	Upgradeable
	Progressing
	Disable
	EvaluationConditionsDetected
	Failing
~~~
https://github.com/openshift/cluster-version-operator/blob/master/install/0000_90_cluster-version-operator_02_servicemonitor.yaml

Author: Sure. 

CR Conditions are not the same as statuses, since the list of values of conditions is determined by Kubernetes. For example https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase.

I believe we should use numbers to represent the statuses in the values, since string values like: Healthy/Warning/Error, can be implemented in many ways that would cause consistency issues and disagreements, like spelling errors, capital letters and preferences of terminology like : `Critical` instead of `Error`. 
Like you, In Kubernetes, I only found examples where status is reported as a label and the value is a boolean.
That is problematic in our case since it would be very hard to control the implementation of each operator and all this effort would not translate into value for the users. 
We need to keep in mind that kubernetes metrics and most metrics in general, are reported from the same project source, which is much simpler then our case, where the metrics would be reported  from each operator separately. So the metric that we add should be very clear and leave no room for confusion.

I chose 3 statuses to allow some room for reporting cases where there are issues in the operator that should be fixed and are related to its health, but the functionality is still in tact(For example, when the number of replicas is not as it should, but there is still a replica that allows the operator to function). 
Also, The existing health metrics, like the ODF/Ceph/NooBaa health metrics also use this methodology.

Reviewer 2: >  list of values of conditions is determined by Kubernetes.

It is determined by the controller author (here the operator author). Conditions is the mechanism used in Kubernetes and largely followed by most of the extensions for surfacing the status of the intent conveyed by a resource.

> That is problematic in our case since it would be very hard to control the implementation of each operator and all this effort would not translate into value for the users.

I am not sure to follow. The approach taken by authors of node, deployment or cluster operator controllers is to provide an extension point to operator authors through the condition label. Ideally they would use common values: "Available", "Progressing" but could add more if they feel that they have a special need. From the user point of view they would start with common values but could also extend if they feel that the special cases foreseen by the operator authors need their attention.

> We need to keep in mind that kubernetes metrics and most metrics in general, are reported from the same project source,

I would argue that it is not the case for cluster operators.

Reviewer 2: @sradco, thanks for that. I like that it is possible to have a single query for all the operators following the naming pattern. I still think that "conditions" are a natural match for things to monitor but it does not have to happen today. Your PR is already a good improvement and if we feel that a finer granularity based on conditions is needed / get a larger adoption we can always extend our recommendation then.
I am looking forward to this PR getting merged when the other reviewers think it is good to go.

Author: The issue is that conditions might not be able to fully indicate the operator health. That is at least the case for our operator. Taking this into account I decided that for now we should focus on the naming and reporting conventions and not the "how to calculate" the metric. I have my thought about that but I think its out of scope for now for this PR. I agree that conditions is defiantly a part of the way to get the operator health status.

 - `Gauge` Value can be increased and decreased as needed.
5. `_total` suffix should be used for accumulating count. If your metrics has labels with high cardinality, like `pod`/`container` it usually means that you can aggregate it more, thus it will not require `_total` suffix.

#### Recommended Metrics

* `operator_health_status` - The metric reports the operator health status. It should be reported by the operator itself based on the internal operator validations.  
  This metric allows the cluster users, to view all operators health statuses with a single metric and improves the overall cluster observability.

>   In order to maintain consistency, Please use the following health statuses as the metric values:

Reviewer: Nit:
```suggestion
  In order to maintain consistency, it is recommended use the following health statuses as the metric values:
```

Author: Updated

  * `0` - Indicates that the operator is healthy and working as expected.
  * `1` - Indicates that the operator has some issues that needs to be addressed and can potentially lead to loss of functionality.
  * `2` - Indicates that the operator is unhealthy and there is a loss of functionality that should be addressed.

  Metric labels:
  * `name` - The operator name.[Required].
  * `reason` - The reason for the health issue, if possible.[Optional].

>   **Note:** In order for this metric to be useful for the users, it MUST be reported in a consistent way across all the operators.  
>   The metric values should be reported as numbers that represent the health statuses, since using strings like: Healthy/Warning/Error, can be implemented in many ways that would cause consistency issues and disagreements, like terminology preferences (For example: `Error` vs. `Unhealthy`), spelling errors, capital letters etc.

Reviewer: ```suggestion
  **Note:** In order for this metric to be useful for the users, it MUST be reported in a consistent way across all the operators. If you would prefer to follow any other level of granularity to indicate health statuses, it should be clearly documented such that an external observer can infer the indicator values accordingly. 
  
It is suggested to report the metric values as numbers that represent the health statuses, since using strings like: Healthy/Warning/Error, can be implemented in many ways that would cause consistency issues and disagreements, like terminology preferences (For example: `Error` vs. `Unhealthy`), spelling errors, capital letters etc.
```

WDYT? This is to ensure that we list out the best practice but not mention that this the only right approach since this doc is upstream. In the downstream documentation we can use the previous version and generalize on the set of indicators based on any further discussion if needed.

Author: Updated

 - `Gauge` Value can be increased and decreased as needed.
5. `_total` suffix should be used for accumulating count. If your metrics has labels with high cardinality, like `pod`/`container` it usually means that you can aggregate it more, thus it will not require `_total` suffix.

#### Recommended Metrics

* `<operator-name-prefix>_operaor_health_status` - The metric reports the operator health status. It should be reported by the operator itself based on the internal operator validations.  

>   This metric allows the cluster users, to view all operators health statuses with a single metric and improves the overall cluster observability.

Reviewer 3: > to view all operators health statuses with a single metric
 
Maybe "to view all operators health statuses with a single query"? As these are different metrics

Author: Updated

  In order to maintain consistency, Please use the following health statuses as the metric values:
  * `0` - Indicates that the operator is healthy and working as expected.
  * `1` - Indicates that the operator has some issues that needs to be addressed and can potentially lead to loss of functionality.
  * `2` - Indicates that the operator is unhealthy and there is a loss of functionality that should be addressed.

  Metric labels:

>   * `name` - The operator name.[Required].

Reviewer 3: Should this be a requirement, as each operator has its prefix?

Author: Updated

 - `Gauge` Value can be increased and decreased as needed.
5. `_total` suffix should be used for accumulating count. If your metrics has labels with high cardinality, like `pod`/`container` it usually means that you can aggregate it more, thus it will not require `_total` suffix.

#### Recommended Metrics

> * `<operator-name-prefix>_operaor_health_status` - The metric reports the operator health status. It should be reported by the operator itself based on the internal operator validations.  

Reviewer 2: there is a typo in the metrics name: `operaor`

Author: Thanks. Updated.

  In order to maintain consistency, it is recommended to use the following health statuses as the metric values:
  * `0` - Indicates that the operator is healthy and working as expected.
  * `1` - Indicates that the operator has some issues that needs to be addressed and can potentially lead to loss of functionality.
  * `2` - Indicates that the operator is unhealthy and there is a loss of functionality that should be addressed.

  Metric labels:
  * `name` - The operator name.[Recommended].
  * `reason` - The reason for the health issue, if possible.[Optional].

  **Query:** To view all operators health statuses with a single query, users can use the following PromQL query, `{__name__=~".*operator_health_status"}`.

  **Note:** In order for this metric to be useful for the users, it MUST be reported in a consistent way across all the operators. If you would prefer to follow any other level of granularity to indicate health statuses, it should be clearly documented such that an external observer can infer the indicator values accordingly.  

>   It is suggested to report the metric values as numbers that represent the health statuses, since using strings like: Healthy/Warning/Error, can be implemented in many ways that would cause consistency issues and disagreements, like terminology preferences (For example: `Error` vs. `Unhealthy`), spelling errors, capital letters etc.

Reviewer 4: This is the kind of thing I was thinking of for why a library could go a long ways here. The library owns the way the metric gets set. Users of the library (i.e. operator authors) would not even need to be concerned at all with what the metric ended up looking like and they may not even need to be concerned with the way we expose that information. In the future, we might also think its a good idea to emit an event. Operator authors would just say "make sure the thing that exposes my operator health is updated to say I am healthy/degraded/failing/unknown/whatever" and the library is the thing that makes sure all operators do whatever stuff on the backend consistently.

Author: @joelanford  At the moment I believe this is a good start, since the library will still be based on this best practice.
We plan to add a monitoring plugin where we can add this as well as part of that effort.

#### Recommended Metrics

* `<operator-name-prefix>_operator_health_status` - The metric reports the operator health status. It should be reported by the operator itself based on the internal operator validations.  
  This metric allows the cluster users, to view all operators health statuses with a single metric and improves the overall cluster observability.

  In order to maintain consistency, it is recommended to use the following health statuses as the metric values:
  * `0` - Indicates that the operator is healthy and working as expected.
  * `1` - Indicates that the operator has some issues that needs to be addressed and can potentially lead to loss of functionality.
  * `2` - Indicates that the operator is unhealthy and there is a loss of functionality that should be addressed.

  Metric labels:
  * `name` - The operator name.[Recommended].

  For example, the Memcached health metric would be:
  memcached_operator_health_status
  labels:
    name: Memcached

  **Query:** To view all operators health statuses with a single query, users can use the following PromQL query, `{__name__=~".*operator_health_status"}`.

>   **Note:** In order for this metric to be useful for the users, it MUST be reported in a consistent way across all the operators. If you would prefer to follow any other level of granularity to indicate health statuses, it should be clearly documented such that an external observer can infer the indicator values accordingly.  

Reviewer 4: ```suggestion
  **Note:** In order for this metric to be useful for the users, it MUST be reported in a consistent way across all the operators. If you would prefer to follow any other level of granularity to indicate health statuses, a different metric name MUST be used, and it should be clearly documented such that an external observer can infer the indicator values accordingly.  
```

Author: Added

 - `Gauge` Value can be increased and decreased as needed.
5. `_total` suffix should be used for accumulating count. If your metrics has labels with high cardinality, like `pod`/`container` it usually means that you can aggregate it more, thus it will not require `_total` suffix.

#### Recommended Metrics

{{% alert title="Warning" color="warning" %}}

> The following operator health metric is new and may be subjected to change 

Reviewer: Could we make this a copy of the note which we have below.

"The metrics below are are the suggested values. If you would prefer to follow any other level of granularity to indicate health statuses, a different metric name MUST be used, and it should be clearly documented such that an external observer can infer the indicator values accordingly."

I do not see a value in just adding that these metrics are subject to change. Imo, it is practically impossible for us to cover all use cases and define a standardized set of metrics that fits all. As mentioned before, as an operator author, I could have "success/failure/critical" as three sets based on my operator's lifecycle, where as someone else could have introduced a new terminology with "success/warning/degraded/critical" etc based on their operator's state, which IMO is not wrong. Though the idea here is to suggest best practices, we can't practically create a standard that serves all the purposes. We could have operator authors suggesting to add "n" number of levels which suite their use case, but it is not practical for us to modify the set based on that. Even if we decide to do so, what is the basis with which we will consider that a suggested health status is useful to be added to this set?

Imo, as experts our task should be to provide necessary docs/helpers to encourage operator authors to set up alerts with an example that defines a generic naming convention and status levels (that is reasonable and less confusing) than us working on defining a specific set with its respective definitions.

> I think a really good way to do that would be to implement something where the author just has to provide a function or hook and then the library does all the heavy lifting to register the metric, and set it to the right 0/1/2 value.

This seems to be generic enough to not only be used in operators but anywhere else. Wrappers like these could be a great addition to prometheus itself than to KB or SDK.

Reviewer 4: @varshaprasad96 I suggested the "subject to change" wording because it seems like it would be good to get some feedback about this from a wider audience before we declare it a best practice. What it seems like we're saying is:
1. All operators should _at least_ specify _this_ metric.
2. We don't have enough feedback to say for sure this specific implementation is a best practice, so we _may_ change some of the specifics (e.g. how to choose the name label, what the various states are, how to convey those states, etc). What we probably will _not_ change is the overall concept of this metric.
3. If you think this metric is not granular enough, there's nothing stopping you from populating a separate metric. Just don't hijack this one.

Maybe we could add a little more to this warning message along the lines of item (2), and maybe add a call to action about providing feedback?

Author: Updated

{{% /alert %}}

> * `<operator-name-prefix>_operator_health_status` - The metric reports the operator health status. It should be reported by the operator itself based on the internal operator validations.  

Reviewer 4: It seems like we should name this simply `operator_health_status` (no operator name prefix)? The operator name will be in the `name` label already so no need to duplicate it here, right?

Author: I investigated this and there is an issue in Prometheus if the `Help` text is reported differently with the same metric name. I believe having a separated metric name for each operator with the same suffix would be the better way here. Also the `name` label will allow for adding  the full and "pretty" name, but its only optional.

Author: See here https://prometheus.io/docs/instrumenting/writing_exporters/#help-strings
```
Prometheus doesn’t like one metric having different help strings. If you’re making one metric from many others, choose one of them to put in the help string.
```

Reviewer 4: This brings up another concern: what happens when we add a fourth enumerated state? We'll have to change the help text to mention that state, we won't be able to ensure that every operator installed on a cluster will update at the same time to use that new help text.

This is another reason to go with a string label for the state enumeration. We can add self-describing state strings without changing the help text.

Author: I don't see a possibility for a fourth enumerated state.
There are 3 states: healthy / lost important functionality / about to loss important functionality(optional)
Having strings adds complications. I dont believe we should encourage developers to add self-describing state strings. As I said before, numbers are much more strict, which is important when you are proposing a metric that each operator needs to implement separatly.

  This metric allows the cluster users, to view all operators health statuses with a single metric and improves the overall cluster observability.

  In order to maintain consistency, it is recommended to use the following health statuses as the metric values:
  * `0` - Indicates that the operator is healthy and working as expected.
  * `1` - Indicates that the operator has some issues that needs to be addressed and can potentially lead to loss of functionality.
  * `2` - Indicates that the operator is unhealthy and there is a loss of functionality that should be addressed.

  Metric Type: `Gauge`

  Metric labels:

>   * `name` - The operator name [Recommended].

Reviewer 4: Maybe add an example so folks are guided to use the same definition of operator name? Also, @fgiloux and I had suggested trying to incorporate a domain aspect somehow to help disambiguate between organizations that implement the same operator concept.

```suggestion
  * `name` - The operator name including domain [Recommended]. (e.g. `yq eval '(.projectName + "_" + .domain) | sub("[\.-]", "_")' <projectRoot>/PROJECT`)
```

Author: This seems too opinionated to me. I believe adding the Memcached example is sufficient here.

  
  Metric Help:
  * Indicates whether the operator health status is healthy (0), warning (1) or critical (2).

  **Example:**  

> ```

Reviewer: ```suggestion
```go
```

Author: Updated

	desc := prometheus.NewGaugeVec(
		prometheus.GaugeOpts{

> 			Name:      "memcached_operator_health_status",
> 			Help:      "Indicates whether the operator health status is healthy (0), warning (1) or critical (2).",
> 		},prometheus.Labels{"name": "Memcached"})

Reviewer 4: ```suggestion
			Name:      "operator_health_status",
			Help:      "Indicates whether the operator health status is healthy (0), warning (1) or critical (2).",
		},prometheus.Labels{"name": "memcached_operator_example_com"})
```

Author: The metric suggests to have the operator name as the prefix and the `operator_health_status` as the suffix. There example is the correct suggestion.

Author: The `name` label in intended for "Pretty" name. Not domain name.
This way the report can display the name and it will be easier for the user to identify the operator

Reviewer 4: Ok, can we change the label name to "prettyName" then to make that obvious? Is that label required or optional?

Author: The name label is recommended, but is optional. I don;t see a reason to name it as "prettyName".
The idea is that in the metric itself the name prefix is short and this label can provide a more detailed name.

```

>   **Query:** To view all operators health statuses with a single query, users can use the following PromQL query, `{__name__=~".*operator_health_status"}`.

Reviewer 4: ```suggestion
  **Query:** To view all operators health statuses with a single query, users can use the following PromQL query, `{__name__="operator_health_status"}`.
```

Author: This is incorrect. The query I wrote is the correct one and it will return all the metrics for all the operators that will report the metric with the `operator_health_status` suffix.

Author: BTW, My first proposal was 1 metric name for all operators. 
But I changed this proposal based on:
1. We recommend in the best practices to add the operator name prefix for all of the operator custom metrics.
This helps to track where the metrics was exposed from which us important when you want to get a better look at the code that generate it. I believe this recommendation is still valid also in this case.
2. Prometheus state that it doesn't handle having different Help text for that same metric. It is reasonable to assume that there would be slight differences in the Help text when the same metric name is used by different opreators.

Reviewer 4: > This helps to track where the metrics was exposed from which is important when you want to get a better look at the code that generated it.

I agree with this, but once again using controller-runtime as an example, the prefix isn't the operator name, it's the library that actually implements the metric. See https://github.com/kubernetes-sigs/controller-runtime/blob/c304e7ec2ee7879f53ce69c12a7aa133853fccdb/pkg/internal/controller/metrics/metrics.go#L25-L64

I would think we'd implement these in operator-lib and use one of these as the prefix:
- `operator_lib_`
- `operator_sdk_`

This piece of the discussion is another reason I think we should go ahead and immediately write a library that implements this proposal, and simultaneously review and comment on that. If we did that, it would immediately bring to the fore all of these gray areas about what actually originates where, who owns what, what can be tweaked and what can't by the operator author, what the cluster admin can rely on, etc.
